### ISSUE TITLE
[security problem] core download signatures should be verified

### ISSUE THREAD

**Issue by yurivict** [libretro/RetroArch#2374]
Currently files available here http://buildbot.libretro.com/nightly/linux/x86_64/latest/ are zipped shared objects. Such approach is susceptible to the MITM attack: attacker can substitute the file on the fly and RetroArch won't even notice this.

See the generic framework for MITM attackers here: https://github.com/infobyte/evilgrade It is only necessary to write a simple plugin to substitute RetroArch core download.

RetroArch should be shipped with its own public key. All cores on the site should be signed with the private key that developers have. OpenSSL can be used to sign files.

**Comment by twinaphex** [comment]
SSL support will be introduced and it is something leiradel will look into.

That being said, it will be an option, it won't be enforced. We cannot accept or tolerate having a hard dependency against any SSL library. If we were talking about very lightweight SSL libraries then enforcing SSL might be an option, but both PolarSSL, OpenSSL and the like don't fit that description, and NIHing it would be quite the totally separate side project.

RA as a project can never ever have hard dependencies against big bloated libraries like this. It will always have to be a 'soft' dependency instead.

**Comment by fr500** [comment]
What if the end user wants to run his own unsigned cores?
Also, OpenSSL is an enormous dependency as twinaphex just said.

I don't think something as OpenSSL can be NIHed reliably.

**Comment by yurivict** [comment]
Users can run their own cores, not problem. But cores downloaded from the central servers should be signed.

**Comment by twinaphex** [comment]
Well, like I said, a hard dependency against any SSL library is not an option, it has to be optional if we are going to do it. Having a hard dependency against any bloated library would be a cardinal sin from our perspective.

Anyway, leiradel expressed interest in implementing SSL, so it will get done, but I don't think we are going to enforce SSL to be used for core / content downloading all the same. It will be optional.

**Comment by yurivict** [comment]
There are other ways to sign: you can use GPG, or you can use some independent implementation.

**Comment by fr500** [comment]
openpgp isn't exactly small either, I agree security is a concern, specially as the project grows but consider this.

Users are using the cores in most cases with software from unverified sources too, and they even download and exchange save states and other data that can be used for exploits too.
Case in point, dosbox, that core (and standalone dosbox too) has full filesystem access without any hacking involved, so signature or not it can be used to wipe an entire filesystem.

The only thing we could achieve with signed cores is plausible deniability and a false sense of security.

**Comment by yurivict** [comment]
> The only thing we could achieve with signed cores is plausible deniability and a false sense of security.

Signing eliminates download process as a suspect. Download via WiFi is the usual way of MITM compromise. There is the device "WiFi Pineapple" (under $100), that allows to systematically hijack and compromise WiFi networks without having any deep knowledge of networking. Same with Kali linux distro. There are communities online that discuss such hacking opportunities on daily basis. You can't just think that "nobody probably will do anything", because they will. Think of the guy sitting in Starbucks for hours with WiFi Pineapple in his bag.

**Comment by fr500** [comment]
I'm acknowledging the risk.
It's still not fixing any security flaws. Hence the false sense of security.

**Comment by twinaphex** [comment]
Let's just see what @leiradel will come up with in cooperation with me later on.

**Comment by RollingStar** [comment]
As a stopgap that's easier to make, how about a separate program that verifies the hashes? Doesn't even have to be part of the RetroArch program. Only thing RetroArch would need to provide is a list of known-good hashes (preferably SHA256 or something else modern, but MD5 is not currently known to be vulnerable to [second pre-image attacks](https://crypto.stackexchange.com/questions/13303/is-md5-second-preimage-resistant-when-used-only-on-fixed-length-messages)).

And to get any part of this started, can the ZIPs be hashed or signed in some way, even if there are no immediate plans to implement hash or signature checking in RetroArch?

**Comment by wastasy** [comment]
Hi,
I would like to know if you have an email address account to send sensitive information in attachments. Can I use libretro@gmail.com?
I look forward to hearing from you.﻿﻿

**Comment by checktext00** [comment]
> openpgp isn't exactly small either, I agree security is a concern, specially as the project grows but consider this.
> 
> Users are using the cores in most cases with software from unverified sources too, and they even download and exchange save states and other data that can be used for exploits too. Case in point, dosbox, that core (and standalone dosbox too) has full filesystem access without any hacking involved, so signature or not it can be used to wipe an entire filesystem.
> 
> The only thing we could achieve with signed cores is plausible deniability and a false sense of security.

This is not true because software can be verified before running it via software lists/xml databases and self-dumping CDs you own. The only thing that is not verifiable currently is the core. It is also downloaded over http by default which makes it a good target to attack users on public wifi networks.
